File: src/main/java/graphics/controls/LevelDiagram.java
package graphics.controls;

import core.circuit.Port;
import core.types.Direction;
import io.Level;
import io.Reader;
import javafx.geometry.Pos;
import javafx.scene.control.Label;
import javafx.scene.layout.Pane;
import javafx.scene.paint.Color;
import javafx.scene.shape.Path;
import javafx.scene.shape.Rectangle;

import static java.lang.Math.*;

/**
 * Created by Mitch on 5/28/2017.
 */
public class LevelDiagram extends Pane {
    private final double OFFSET = 0.15;
    private final double ARROW_ANGLE  = PI*0.65;
    private final double ARROW_LENGTH = 0.13;

    private double size;

    public LevelDiagram(double size){
        this.size = size;
        this.setWidth(size);
        this.setHeight(size);
    }

    public void set_level(Level level){
        this.getChildren().clear();
        double offset = OFFSET * size;
        Rectangle sides  = new Rectangle(0, 0, size, size);
        sides.setFill(Color.ANTIQUEWHITE);
        Path square = HandDrawn.rectangle(offset, offset, size - offset, size - offset);
        this.getChildren().addAll(sides, square);

        for (int i = 0; i < 4; i++){
            Port port = Port.create(level.inputs()[i], level.outputs()[i]);
            if(port.port_type().id() < 2){
                Direction arrow_rot = Direction.apply(port.is_output() ? i : i+2);
                Path path = arrow(Direction.apply(i), arrow_rot);
                Label label = number(Direction.apply(i), port.capacity());

                this.getChildren().addAll(path, label);
            }
        }
    }

    private Path arrow(Direction side, Direction arrow_rotation){
        final double[][] ARROW_POSITIONS = {
                {0.5,           OFFSET},
                {1.0 - OFFSET,  0.5},
                {0.5,           1.0 - OFFSET},
                {OFFSET,        0.5}
        };
        double a = OFFSET * size;
        double p = OFFSET / (2*tan(ARROW_ANGLE/2)) * size;
        double c = cos(ARROW_ANGLE/2)*ARROW_LENGTH * size;
        double s = sin(ARROW_ANGLE/2)*ARROW_LENGTH * size;

        Path[] lines = {
            HandDrawn.line( a*.5, -a+p,  a*.5, a),
            HandDrawn.line(-a*.5, -a+p, -a*.5, a),
            HandDrawn.line(0, -a,  s, -a+c),
            HandDrawn.line(0, -a, -s, -a+c)
        };

        Path path = new Path();
        for (Path line : lines)
            path.getElements().addAll(line.getElements());

        path.setRotate(arrow_rotation.n() * 90);
        path.setLayoutX(ARROW_POSITIONS[side.n()][0] * size);
        path.setLayoutY(ARROW_POSITIONS[side.n()][1] * size);

        return path;
    }

    private Label number(Direction side, int capacity){
        final double[][] LABEL_POSITIONS = {
                {0.5-1.5*OFFSET,    0.0},
                {1.0-OFFSET,        0.5-1.5*OFFSET},
                {0.5-1.5*OFFSET,    1.0 - OFFSET},
                {0.0,               0.5-1.5*OFFSET}
        };
        final double FONT_OFFSET = 0.05; //Change this after changing font size
        double a = OFFSET * size;

        Label l = new Label(capacity+"");
        l.getStyleClass().add("capacity_marker");
        l.setPrefWidth(a);
        l.setPrefHeight(a);
        l.setAlignment(side.n() == 1 ? Pos.CENTER_LEFT : Pos.CENTER_RIGHT);

        l.setLayoutX(LABEL_POSITIONS[side.n()][0] * size);
        l.setLayoutY((LABEL_POSITIONS[side.n()][1] - FONT_OFFSET) * size);
        return l;
    }

}

File: src/main/scala/graphical/controls/LevelDiagram.scala
package main.scala.graphical.controls

import scala.math._
import core.circuit.Port
import core.circuit.Port.PortType
import core.types.Direction
import io.Level

import scalafx.scene.control.Label
import scalafx.scene.layout.Pane
import scalafx.scene.shape.{Path, PathElement, Rectangle}
import scalafx.geometry.Pos
import scalafx.scene.Node
import scalafx.scene.paint.Color

/**
  * Created by Mitch on 7/30/2017.
  */
class LevelDiagram(size: Double) extends Pane {
  private val OFFSET = 0.15
  private val ARROW_ANGLE = Pi * 0.65
  private val ARROW_LENGTH = 0.13
  private val root = new Pane()

  children = root
  prefWidth  = size
  prefHeight = size

  def set_level(level: Level): Unit = {
    def make_label(d: Direction): Seq[Node] = {
      val port = Port.create(level.inputs(d), level.outputs(d))
      if (port.port_type != PortType.UNUSED) {
        val arrow_rot = Direction(if (port.is_output) d else d+2)
        Seq (arrow(Direction(d), arrow_rot),
             number(Direction(d), port.capacity)
        )
      } else Seq.empty[Node]
    }

    val offset = OFFSET * size
    this.root.children = Seq(
        new Rectangle {x = 0; y = 0; width = size; height = size; fill = Color.AntiqueWhite},
        new Path {elements = HandDrawn.rectangle(offset, offset, size - offset, size - offset)}
      ) ++ Direction.values.flatMap(make_label)
  }


  private def arrow(side: Direction, arrow_rotation: Direction): Path = {
    val ARROW_POSITIONS = Seq(
      Seq(0.5,          OFFSET),
      Seq(1.0 - OFFSET, 0.5),
      Seq(0.5,          1.0 - OFFSET),
      Seq(OFFSET,       0.5)
    )
    val a = OFFSET * size
    val p = OFFSET / (2 * tan(ARROW_ANGLE / 2)) * size
    val c = cos(ARROW_ANGLE / 2) * ARROW_LENGTH * size
    val s = sin(ARROW_ANGLE / 2) * ARROW_LENGTH * size

    new Path {
      elements = Seq(
        HandDrawn.line(a*.5,  -a+p, a*.5,   a),
        HandDrawn.line(-a*.5, -a+p, -a*.5,  a),
        HandDrawn.line(0,     -a,   s,   -a+c),
        HandDrawn.line(0,     -a,   -s,  -a+c)
      ).flatten
      rotate = arrow_rotation * 90.0
      layoutX = ARROW_POSITIONS(side)(0) * size
      layoutY = ARROW_POSITIONS(side)(1) * size
    }
  }

  private def number(side: Direction, capacity: Int): Label = {
    val LABEL_POSITIONS = Seq (
      Seq(0.5 - 1.5 * OFFSET, 0.0),
      Seq(1.0 - OFFSET, 0.5 - 1.5 * OFFSET),
      Seq(0.5 - 1.5 * OFFSET, 1.0 - OFFSET),
      Seq(0.0, 0.5 - 1.5 * OFFSET)
    )
    val FONT_OFFSET = 0.02 //Change this after changing font size
    val a = OFFSET * size

    new Label {
      text = s"$capacity"
      styleClass = Seq("capacity_marker")
      prefWidth  = a
      prefHeight = a
      alignment = if (side.n == 1) Pos.CenterLeft else Pos.CenterRight
      layoutX = LABEL_POSITIONS(side)(0) * size
      layoutY = (LABEL_POSITIONS(side)(1) - FONT_OFFSET) * size
    }
  }
}















